src/effects/native/autopaneffect.cpp

```diff
@@ -0,0 +1,192 @@
+    delay->setDescription("Controls length of the delay");
+    delay->setControlHint(EffectManifestParameter::CONTROL_KNOB_LINEAR);
+    delay->setSemanticHint(EffectManifestParameter::SEMANTIC_UNKNOWN);
+    delay->setUnitsHint(EffectManifestParameter::UNITS_UNKNOWN);
+    delay->setMinimum(0.0);
+    delay->setMaximum(0.02);    // 0.02 * sampleRate => 20ms
+    delay->setDefault(0.01);
+    
+    return manifest;
+}
+
+AutoPanEffect::AutoPanEffect(EngineEffect* pEffect, const EffectManifest& manifest)
+        : 
+          m_pSmoothingParameter(pEffect->getParameterById("smoothing")),
+          m_pPeriodParameter(pEffect->getParameterById("period")),
+          m_pDelayParameter(pEffect->getParameterById("delay"))
+           {
+    Q_UNUSED(manifest);
+}
+
+AutoPanEffect::~AutoPanEffect() {
+}
+
+void AutoPanEffect::processChannel(const ChannelHandle& handle, PanGroupState* pGroupState,
+                              const CSAMPLE* pInput,
+                              CSAMPLE* pOutput, const unsigned int numSamples,
+                              const unsigned int sampleRate,
+                              const EffectProcessor::EnableState enableState,
+                              const GroupFeatureState& groupFeatures) {
+    Q_UNUSED(handle);
+    
+    PanGroupState& gs = *pGroupState;
+    
+    if (enableState == EffectProcessor::DISABLED) {
+        return;
+    }
+    
+    CSAMPLE period = m_pPeriodParameter->value();
+    if (groupFeatures.has_beat_length) {
+        // 1/8, 1/4, 1/2, 1, 2, 4, 8, 16, 32, 64
+        double beats = pow(2, floor(period * 9 / m_pPeriodParameter->maximum()) - 3);
+        period = groupFeatures.beat_length * beats;
+    } else {
+        // max period is 50 seconds
+        period *= sampleRate * 25.0;
+    }
+    
+    CSAMPLE stepFrac = m_pSmoothingParameter->value();
+    
+    if (gs.time > period || enableState == EffectProcessor::ENABLING) {
+        gs.time = 0;
+    }
+    
+    // Normally, the position goes from 0 to 1 linearly. Here we make steps at
+    // 0.25 and 0.75 to have the sound fully on the right or fully on the left.
+    // At the end, the "position" value can describe a sinusoid or a square
+    // curve depending of the size of those steps.
+    
+    // coef of the slope
+    // a = (y2 - y1) / (x2 - x1)
+    //       1  / ( 1 - 2 * stepfrac)
+    float a = stepFrac != 0.5f ? 1.0f / (1.0f - stepFrac * 2.0f) : 1.0f;
+    
+    // define the increasing of gain (the music sounds lower if only one 
+    // channel is enabled).
+    // TODO(jclaveau) : Challenge this value
+    float lawCoef = 1.0f / sqrtf(2.0f) * 2.0f;
+    
+    // size of a segment of slope (controled by the "strength" parameter)
+    float u = (0.5f - stepFrac) / 2.0f;
+    
+    gs.frac.setRampingThreshold(positionRampingThreshold);
+    gs.frac.ramped = false;     // just for debug
+    
+    // prepare the delay buffer
+    float delay = round(m_pDelayParameter->value() * sampleRate);
```

Comment:
The parameter is only here to challenge the delay value, I totally agree with you.
I think the offset can have a fixed value too (probably changing with the period) and finally the width parameter could have a dedicated knob (i'm speacking about the parameters provided in the SoundToys PanMan plugin for Protools).